beluga-aws-sdk/src/lib.rs

```diff
@@ -17,45 +20,75 @@ impl TunnelManager {
     ///
     /// This function creates a new [`TunnelManager`] instance by subscribing to
     /// the MQTT topic `$aws/things/{thing_name}/tunnels/notify` with a quality
-    /// of service level of at least once. The resulting
-    /// [`MqttSubscriber`](Subscriber) is stored in the [`TunnelManager`]
-    /// instance.
+    /// of service level of at least once.
     ///
     /// # Arguments
     ///
     /// * `mqtt` - The `MqttClient` instance to use for the MQTT subscription.
     /// * `thing_name` - The name of the AWS IoT thing to subscribe to.
     pub async fn new(mqtt: MqttClient, thing_name: &str) -> Result<Self> {
-        let subscriber = mqtt
+        let mut subscriber = mqtt
             .subscribe(
                 format!("$aws/things/{thing_name}/tunnels/notify").as_str(),
                 QoS::AtLeastOnce,
             )
             .await?;
 
-        Ok(Self { subscriber })
+        let cancel = CancellationToken::new();
```

Comment:
The `new` function should take this as an argument.

Reply:
Maybe let's have a shutdown method? What do you think?

Reply:
I think it only really makes sense to use cancellation tokens if we expose them to the user. The benefit then becomes that you can have a way to basically instantly propagate shutdown through the entire app like this:

```rust
async fn main() {
  let cancel = CancellationToken::new();

  // since we're passing down child tokens instead of clones these two can never accidentally shut down the entire app.
  let tunnel_manager = TunnelManager::new(..., cancel.child_token()).await.unwrap();
  let some_other_service = OtherService::new(cancel.child_token());

  // spawn our shutdown listener
  tokio::spawn(async move {
    tokio::signal::ctrl_c().await?;
    cancel.cancel();
    Ok(())
  });

  cancel.cancelled().await;

  // at this point all services from above have already started shutting down (in parallel)!
  // we don't need to tell each one to do so explicitly, we can just wait for them to finish.

  tunnel_manager.shutdown().await.unwrap();
  some_other_service.shutdown().await.unwrap();

  // fully shut-down
}
```

If you don't want that that's also completely fine, but then we don't need the overhead of a cancellation token and we should just use an `mpsc` (pass the senders to the task and drop the receiver to shut down).

Reply:
Actually wait... I just learned about <https://docs.rs/tokio-util/0.7.10/tokio_util/task/task_tracker/struct.TaskTracker.html>!
I completely missed that until now because it was only added in the latest version.
That sounds perfect for our use case, what do you think?

Reply:
> Actually wait... I just learned about https://docs.rs/tokio-util/0.7.10/tokio_util/task/task_tracker/struct.TaskTracker.html! I completely missed that until now because it was only added in the latest version. That sounds perfect for our use case, what do you think?

looks awesome, let's use it.